### PR TITLE
fix #12971: note collisions in playback

### LINKED ISSUE #12971: [MU4 Issue] Falls & other bends do not play on last note of a tie
**Describe the bug**
The new fall playback is basically good, but fails if adding to the end of a tied note.

**To Reproduce**
1. Score for trumpet
2. Enter two tied notes
3. Attach a fall to the second
4. Play
5. See error: the fall does not play

**Expected behavior**
Fall should play

**Screenshots**
image

**Platform information**
 - OS: Linux

**Additional context**
It works if you add to the first note of the tie, although that's incorrect notation

### PR DESCRIPTION
**Pull request by mirabilos**
MIDI export will still have colliding note if they are in
different staves; same stave is fixed though

### CODE REVIEW

**Comment by MarcSabatella** [comment]
I haven't really analyzed the logic of your fixupMIDI() function, but it appears it would take only O(n) time and not O(n^2), which is good.  And in any case it shouldn't affect overall score responsiveness, just the delay when hitting the play button.  It should be tested with a reasonably large score. to be sure it isn't too bad.  But I think your overall approach is sound.

**Comment by Jojo-Schmitz** [comment]
There are a couple MIDI tests failing in Travis CI. Well possible that the references need fixing, not the code

File: synthesizer/event.cpp [hunk 1]
```diff
@@ -370,5 +370,43 @@ void EventList::insert(const Event& e)
...
+      auto it = begin();
+
+      while (it != end()) {
+            bool discard = false;
+
+            /* ME_NOTEOFF is never emitted, no need to check for it */
+            if (it->second.type() == ME_NOTEON) {
+                  unsigned short np = nowPlaying[it->second.channel()][it->second.pitch()];
+                  if (it->second.velo() == 0) {
+                        if (np == 0) {
```
Comment (Jojo-Schmitz) [inline synthesizer/event.cpp]: Too many unneeded braces for my taste

Comment (mirabilos) [reply synthesizer/event.cpp]: OK. Some people at work insist on them, but I agree and will remove them (and see after the tests).

Comment (Jojo-Schmitz) [reply synthesizer/event.cpp]: And maybe using `if ((np == 0) || (--np > 0))` here? Not sure though whether this might undefined behavoir

Comment (Jojo-Schmitz) [reply synthesizer/event.cpp]: as per https://en.wikipedia.org/wiki/Sequence_point this should be clean, the `||` is a sequence point.Leave a comment

Comment (mirabilos) [reply synthesizer/event.cpp]: Hmm, okay. I had a few debugging `fprintf(stderr, …);` in between, which is why I had the separate logic.

**Comment by mirabilos** [comment]
@Jojo-Schmitz style concerns addressed, tests fixed, commits squashed

File: synthesizer/event.h [hunk 2]
```diff
@@ -311,7 +311,12 @@ class EventList : public QList<Event> {
       void insertNote(int channel, Note*);
       };
 
-class EventMap : public std::multimap<int, NPlayEvent> {};
+class EventMap : public std::multimap<int, NPlayEvent> {
+      int _highestChannel = 15;
+   public:
+      void fixupMIDI();
+      void haveChannel(int c) { if (c > _highestChannel) _highestChannel = c; }
```
[the remark this answers is not in the document]
Comment (Jojo-Schmitz) [reply synthesizer/event.h]: Agreed, I've stumbled across this too

**Comment by mirabilos** [comment]
Joachim Schmitz dixit:

>+      void haveChannel(int c) { if (c > _highestChannel) _highestChannel = c; }
>
>Agreed, I've stumbled across this too

How about registerChannel() then?

The idea is that this is called for all channels anyway.
In the current implementation we only need the highest
number, since we use a non-sparse array, but like this,
nobody gets the idea to optimise prematurely.

**Comment by mirabilos** [comment]
@Jojo-Schmitz @anatoly-os renamed

**Comment by lasconic** [comment]
Let's focus on https://github.com/musescore/MuseScore/pull/3514 since it contains this PR.

**Pull request closed without merging**